tests/framework/testutils/log_observer.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testutils
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	zapobserver "go.uber.org/zap/zaptest/observer"
+)
+
+type LogObserver struct {
+	ob  *zapobserver.ObservedLogs
+	enc zapcore.Encoder
+
+	mu      sync.Mutex
+	entries []zapobserver.LoggedEntry
+}
+
+func NewLogObserver(level zapcore.LevelEnabler) (zapcore.Core, *LogObserver) {
+	// align with zaptest
+	enc := zapcore.NewConsoleEncoder(zap.NewDevelopmentEncoderConfig())
+
+	co, ob := zapobserver.New(level)
+	return co, &LogObserver{
+		ob:  ob,
+		enc: enc,
+	}
+}
+
+// Expect returns the first N lines containing the given string.
+func (logOb *LogObserver) Expect(ctx context.Context, s string, count int) ([]string, error) {
+	return logOb.ExpectFunc(ctx, func(log string) bool { return strings.Contains(log, s) }, count)
+}
+
+// ExpectFunc returns the first N line satisfying the function f.
+func (logOb *LogObserver) ExpectFunc(ctx context.Context, filter func(string) bool, count int) ([]string, error) {
+	i := 0
+	res := make([]string, 0, count)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		entries, total := logOb.syncLogs()
+
+		for ; i < total; i++ {
```

Comment:
Hi, `i` is increasing item and we don't need to scan the entities which has been scanned. So I declare the `i` before entering `loop`.

Reply:
Please see my comment above `Note the ob.TakeAll() (line 101) truncates all observed logs.`. It means each time `logOb.syncLogs()` (line 65) will get different log slice, so my understanding is that you should reset i to 0.  Does it make sense?

Reply:
`syncLogs` is to dump the logged entries from `zapobserver` and truncates all the entries in `zapobserver` side. It's like `move` action.

For example, 

| Sync Time  |  LogObserver entries  |  zapobserver entries |
| ----- | --- | --- |
|  | [] | [A, B ] |
| T1 | [A, B] | [] |
|  | [A, B] | [C, D, F] |
| T2 | [A, B, C, D, F] | [] |

The `LogObserver` always contains all logged entries after synced. Even if the `entries`'s underlaying slice(memory address) has been changed by go-runtime `append`, both the order and the content are the same. So I think we can reuse the `i` to continue expect action. Is it acceptable?

Reply:
Got it. But I don't understand why you cache all entries? Can we change line 101 to something like below?

```
return logOb.ob.TakeAll()
```

Also why do you return the two values in `(*LogObserver) syncLogs()`, users can easily get the count.

Reply:
It seems that If we monitor/analyze the log multiple times, then we still need to cache all log entries.

Reply:
> Also why do you return two values in (*LogObserver) syncLogs(), users can easily get the count.

`append` might grow the slice's capacity and then next `append` might not copy. Since I don't use lock when I scan the entries, I think it's better to should return slice with len so that there is no race issue. I will add a comment to the function.

> It seems that If we monitor/analyze the log multiple times, then we still need to cache all log entries.

Yes. It allows caller to expect multiple times. It is aligned with `pkg/expect` behavior.

Reply:
> Since I don't use lock when I scan the entries, I think it's better to should return slice with len so that there is no race issue.

It isn't correct to me. No matter how you update/append the `logOb.entries`, it will not affect the already returned slice(entries at line 65). Note that in golang, a slice is actually a struct (with 3 fields), it's returned by value.

Reply:
You're correct. The slice header maintains the `Len` even if they shares the same underlaying `Data`. So it's fine to return slice only. I will update it. Thanks for pointing it out.

Reply:
updated! PTAL, thanks